### ISSUE TITLE
Should tobytes work on object arrays

### ISSUE THREAD

**Issue by jakirkham** [numpy/numpy#12476]
Currently one can create an `object`-type `ndarray` and coerced to `bytes`. However what one gets is effectively the pointers from that `object` array. It's unclear how useful this is. Should it be allowed or should it error?

FWIW `memoryview`s also support this same behavior the same way. Though it is equally questionable in that case whether they are useful either.

Thoughts?

### Reproducing code example:

```python
In [1]: import numpy as np

In [2]: a = np.asarray([None])

In [3]: a.tobytes()
Out[3]: b'\xf8\xc8\xb3\x06\x01\x00\x00\x00'
```

### Numpy/Python version information:

```python
In [1]: import sys, numpy; print(numpy.__version__, sys.version)
1.15.1 3.6.6 | packaged by conda-forge | (default, Jul 26 2018, 09:55:02) 
[GCC 4.2.1 Compatible Apple LLVM 6.1.0 (clang-602.0.53)]
```

**Comment by eric-wieser** [comment]
I don't think this is harmful, although I suppose it's surprising given that `a.view(np.uint8)` is forbidden. The later makes sense to restrict though, as allowing it would allow rewriting of the pointers.

**Comment by jakirkham** [comment]
Hmm...I'm not sure. What value is `tobytes` providing here actually?

Are we sure it couldn't be misused or abused? For example serializing an array?

That's interesting. FWIW `memoryview`s have no problem casting this to `uint8`.

```python
In [1]: import numpy as np

In [2]: a = np.asarray([None])

In [3]: np.asarray(memoryview(a).cast('B'))
Out[3]: array([248, 136,  55,  13,   1,   0,   0,   0], dtype=uint8)
```

**Comment by seberg** [comment]
Not much value maybe, but I think the unsafe function would be `np.ndarray` (or whatever you use, if there is some other reasonable way) and not the `tobytes`.

**Comment by BartMassey** [comment]
The real problem comes here:

```Python
>>> numpy.array(None).tobytes()
b'\x10*\x90\x00\x00\x00\x00\x00'
```

Given that `None` is a valid scalar object, that would be hard to avoid. But it's almost certainly not what anyone would want. (Came here because I just got bit by this — really extended my debugging time.)

Best would be to fail to turn `None` into bytes; second-best would be to turn it into an empty byte array. Turning it into a byte representation of its pointer is no good.

**Comment by BartMassey** [comment]
After thinking about this, I guess my preference would be for `tobytes()` to only work on user objects if they implement the `__bytes__()` method, at which point this would be used to serialize them. This would eliminate the most common bugs while still allowing custom number objects to be serializable.

**Comment by eric-wieser** [comment]
Such a behavior would not be consistent with the current documentation, which says

> Construct Python bytes containing the raw data bytes in the array.
>
> Constructs Python bytes showing a copy of the raw contents of data memory

**Comment by eric-wieser** [comment]
At the very least we should stick a warning in the docs, and possibly emit a UserWarning.

**Comment by BartMassey** [comment]
Thanks for the quick response!

**Comment by rossbar** [comment]
This was discussed during the triage meeting - the consensus seemed to be that raising a UserWarning here would be reasonable (and updating the docs is uncontroversial).
